Automattic/woocommerce-payments
# Add support for subscriptions admin payment method changes

Fixes #700 

#### Changes proposed in this Pull Request

* Add support for subscription_payment_method_change_admin
* Display the shopper's tokens in a select element when editing a subscription
* ~Display user saved payment methods at the user view and edit page~ (moved to #900 )

#### Testing instructions



These flows need to be tested using the `subscription-edit-page.js`, which is applied when using Subscriptions `<= 3.0.7`, and using the new `payment_meta_input_` action, which needs a Subscriptions version based on subscriptions#3798 and its version manually set to `3.0.8`or higher.

The error flows are more to check the validation of the metadata, as they're unlikely to happen during real usage.

1. **Succesful update**
    1.1. On the subscription edit page, click on the edit button for its billing details
    1.2. Scroll down until "Saved payment method", assert that the dropdown contains all the saved cards for this customer
    1.3. Select another card and update the subscription
    1.4. Make a renewal and assert that the selected card was used
2. **No token selected**
    2.1. On the billing details edit page, use the JS console to set the select value to an invalid one 
`document.getElementById( '_payment_method_meta[woocommerce_payments][wc_order_tokens][token]' ).value = 0`
    2.2. Click on update and assert that the `"A customer saved payment method was not selected for this order"` error pops up
3. **Invalid token**
    3.1. On the billing details edit page, use the JS console to create a new option for another user token 
    ```
    var opt = document.createElement( 'option' );
    opt.value = 584954;
    document.getElementById( '_payment_method_meta[woocommerce_payments][wc_order_tokens][token]' ).appendChild( opt );
    ```
    3.2. Select the new option and click on update
    3.3. Assert that the `"The saved payment method selected is invalid or does not exist"` error pops up
4. **Token from another customer**
    4.1. On the billing details edit page, use the JS console to create a new option for another user token 
    ```
    var opt = document.createElement( 'option' );
    opt.value = 145; // Change this for another user token ID
    document.getElementById( '_payment_method_meta[woocommerce_payments][wc_order_tokens][token]' ).appendChild( opt );
    ```
    4.2. Select the new option and click on update
    4.3. Assert that the `"The saved payment method selected does not belong to this order's customer"` error pops up

-------------------
